konnectivity-client/pkg/client/client.go

```diff
@@ -66,15 +66,15 @@ var _ clientConn = &grpc.ClientConn{}
 // gRPC based proxy service.
 // Currently, a single tunnel supports a single connection, and the tunnel is closed when the connection is terminated
 // The Dial() method of the returned tunnel should only be called once
-func CreateSingleUseGrpcTunnel(address string, opts ...grpc.DialOption) (Tunnel, error) {
-	c, err := grpc.Dial(address, opts...)
+func CreateSingleUseGrpcTunnel(ctx context.Context, address string, opts ...grpc.DialOption) (Tunnel, error) {
```

Comment:
We should add documentation/warning about the context on the tunnel. We should *NOT* be using request level context if there is any chance that the tunnel will be (re)used for more than 1 request.

Reply:
to be fair, the "SingleUse" name and "The Dial() method of the returned tunnel should only be called once" docs are already pretty blunt

Reply:
That's fair but I think slightly different than what I'm getting at. It is possible to send (serially) multiple requests over a connection returned by the dial/tunnel. We should not be using a request level context on the create tunnel request as it could cause the tunnel to terminate early.

Reply:
if the tunnel is being created just for one request (is it?) then we want the dial for establishing the tunnel to be bounded to that request's lifetime, right?

Reply:
Yes. However we seem to have a single place from which this is called in K/K (https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go#L206). I cannot convince myself that the tunnel is only used for a single request. In fact the hung log request/tunnel scenario seemed to suggest that tunnels were some times reused.

Reply:
I also have a hard time tracing the kube-apiserver callsites. (oof: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L224)

For this PR: In the worst case, we can update K/K with context.TODO()